### DISCUSSION TITLE
How do I apply updated sieve script without restarting the container?

### DISCUSSION THREAD

**Discussion by WantToLearnJapanese** [docker-mailserver/docker-mailserver#2633] [Q&A] [answered]
How do I apply the updated sieve script without restarting the container?

**Comment by casperklein** [comment, marked as answer]
I guess dovecot have to be restarted? You can restart individual services instead of the whole container by using: `docker exec mailserver supervisorctl restart dovecot`

**Reply by WantToLearnJapanese** [reply]
> I guess dovecot has to be restarted? You can restart individual services instead of the whole container by using: `docker exec mailserver supervisorctl restart dovecot`

Sadly that command seems not working. I am writing on before.dovecot.sieve for globally applying the script. But, while restarting the container make the edited script work, `docker exec mailserver supervisorctl restart dovecot` does not.

**Reply by casperklein** [reply]
I am not using sieve myself.. One last try: `docker exec mailserver sievec /tmp/docker-mailserver/before.dovecot.sieve`

**Reply by Waldorf3** [reply]
> docker exec mailserver sievec /tmp/docker-mailserver/before.dovecot.sieve

Running this command generates or updates a before.dovecot.svbin in the same folder, but doesn't otherwise activate the chage. The file doesnt seem to be used, or at least not necessary, as down/up the stack does not create a .svbin file.

> docker exec mailserver supervisorctl restart dovecot

Does not appear to have any effect.

The only way I make it work is by stoping/starting (down/up) the stack. Not even a restart activates it. 

I'd appreciate guidance to how it may be possible to activate changed sieve filters without stopping the mailserver. 

I'm surprised there is so little information available about sieve on the internet/youtube in general.  I'm a degoogled/gmail refugee, and the filtering system was the thing I found most useful in Gmail. I am really surprised not more people seem to be using it.

**Reply by polarathene** [reply]
> but doesn't otherwise activate the change. The file doesnt seem to be used, or at least not necessary

You can look up our start up script for handling this, we use the config volume location to copy the file internally, and then run `sievec` on the internal location:

https://github.com/docker-mailserver/docker-mailserver/blob/02415b03dce763a717747bb737e4698ddcd7aefd/target/scripts/startup/setup.d/dovecot.sh#L76-L81

That method then runs some additional operations, only the `chown` seems to be applied to this `before.dovecot.svbin` file, ensuring `dovecot:root` ownership (_probably because ownership would have been `root:root` as the `sievec` command was run as `root`_):

https://github.com/docker-mailserver/docker-mailserver/blob/02415b03dce763a717747bb737e4698ddcd7aefd/target/scripts/startup/setup.d/dovecot.sh#L89-L91

There's also separate logic here that is to ensure the `.svbin` files have a newer mtime value than their source `.sieve` files:

https://github.com/docker-mailserver/docker-mailserver/blob/02415b03dce763a717747bb737e4698ddcd7aefd/target/scripts/startup/setup-stack.sh#L26-L35

There's [handling elsewhere for other `.sieve` files](https://github.com/search?q=repo%3Adocker-mailserver%2Fdocker-mailserver%20.sieve&type=code) supported in DMS should that be relevant to know too.

---

> The only way I make it work is by stoping/starting (down/up) the stack. Not even a restart activates it.

DMS skips a bunch of initial setup scripts when the container is restarted (_avoids various issues that can happen_):

https://github.com/docker-mailserver/docker-mailserver/blob/02415b03dce763a717747bb737e4698ddcd7aefd/target/scripts/start-mailserver.sh#L177-L200

You need to use `docker compose down && docker compose up` or `docker compose up --force-recreate` so that the previous DMS instance is destroyed (_the containers internal state, has no affect on your bind mount volumes_).

---

> I'd appreciate guidance to how it may be possible to activate changed sieve filters without stopping the mailserver.

You can reference the commands above for what needs to be done.

This was brought up in [another discussion earlier this year](https://github.com/orgs/docker-mailserver/discussions/3797#discussioncomment-8430681), with the option to automate this via the `check-for-changes.sh` service that would detect updates to files and run the necessary commands for you.

To add that support, the path to monitor on the config volume needs to be added here:

https://github.com/docker-mailserver/docker-mailserver/blob/02415b03dce763a717747bb737e4698ddcd7aefd/target/scripts/helpers/change-detection.sh#L30-L42

Followed by a line like this one:

https://github.com/docker-mailserver/docker-mailserver/blob/02415b03dce763a717747bb737e4698ddcd7aefd/target/scripts/check-for-changes.sh#L141

So something like this should work:

```bash
[[ ${CHANGED} =~ ${DMS_DIR}/before.dovecot.sieve ]] && _run_sievec \
  "${DMS_DIR}/before.dovecot.sieve" \
  '/usr/lib/dovecot/sieve-global/before/50-before.dovecot.sieve'

[[ ${CHANGED} =~ ${DMS_DIR}/after.dovecot.sieve ]] && _run_sievec \
  "${DMS_DIR}/after.dovecot.sieve" \
  '/usr/lib/dovecot/sieve-global/after/50-after.dovecot.sieve'
```

And add this function snippet into the `check-for-changes.sh` file for the above snippet to call:

```bash
function _run_sievec() {
  local EXTERNAL_SIEVE=${1}
  local INTERNAL_SIEVE=${2}
  local INTERNAL_SVBIN="${INTERNAL_SIEVE%.*}.svbin"

  # Make an internal copy and compile the sieve script:
  install --mode 755 --owner dovecot --group root "${EXTERNAL_SIEVE}" "${INTERNAL_SIEVE}"
  sievec "${INTERNAL_SIEVE}"
  chown dovecot:root "${INTERNAL_SVBIN}"
  
  # Ensure the mtime of the svbin is newer than the input:
  # https://github.com/docker-mailserver/docker-mailserver/pull/3779
  touch -d '2 seconds ago' "${INTERNAL_SIEVE}"
  touch -d '1 seconds ago' "${INTERNAL_SVBIN}"
}
```

The `install` command is handling `cp` + `chown` + `chmod`, but the `.svbin` created with `sievec` after is going to need at least the `chown` applied on it as well I think.

If you don't want to modify `check-for-changes.sh` that `_run_sievec` function should be all you need (_`user-patches.sh` isn't viable for container restarts, you'd need to make your own script or manually run the individual commands_). Just pass it the external (DMS config volume) and internal paths like shown in the prior snippet and you should be good 👍 (_you may need to also run `dovecot reload` afterwards, the `check-for-changes.sh` approach would do this implicitly_)

---

> I'm surprised there is so little information available about sieve on the internet/youtube in general.

You can search DMS issues and discussions (_remove the default `open` filter_) for keywords like `sieve` and find various information/tips.

Our docs should have some useful information on the topic too (_**EDIT:** Until v15 is released, you'd want to view the `:edge` version of the docs where I updated the sieve page_).

**Reply by Waldorf3** [reply]
> This was brought up in [another discussion earlier this year](https://github.com/orgs/docker-mailserver/discussions/3797#discussioncomment-8430681), with the option to automate this via the check-for-changes.sh service that would detect updates to files and run the necessary commands for you.

Ah, sorry. As you might have guessed from my avatar, I am a senior citizen, I forgot.  

In any case, I'm not a developer and the above is way above my paygrade.  For the restart I will just proceed to up/down the stack (i.e. the docker-compose) until a better way may be included some day.

Bonus question. Does this implemenentation support "extlists" ?

![image](https://github.com/user-attachments/assets/2a9a55b5-ff2d-4018-8f6b-a6524d970e9b)

**Reply by polarathene** [reply]
> In any case, I'm not a developer and the above is way above my paygrade.

No worries, it's added to my [public backlog of smaller tasks](https://github.com/docker-mailserver/docker-mailserver/issues/4148) that someone else may get around to tackling before I do 😎 (_most of the work is covered above, just needs some verification and perhaps a test, ideally it'd refactor the existing related scripts too_)

---

> Bonus question. Does this implementation support "extlists" ?

Is your screenshot from the Dovecot docs? I found [this](https://github.com/dovecot/pigeonhole/blame/398501d39179545b935394be8b9ba0123ad191ab/README#L168) which seems to be about the Dovecot sieve implementation which states that `extlists` is planned since 2016, but maybe that is outdated?

I did not have much luck finding anything else that says Dovecot supports the extension. AFAIK the Dovecot docs list what sieve extensions are supported by default and the others that they offer which needs the `require` import to use them. I'm not sure if there's a way in DMS to query what the available sieve extensions available are though.

**Reply by Waldorf3** [reply]
> Is your screenshot from the Dovecot docs?

It's from ChatGPT ;)
